### PR TITLE
Made Anonymous Closures for functions passed and called as a param

### PR DESCRIPTION
**Pull request by torch2424**
This is still a work in progress, cc @DuncanUszkay1 :smile: 

But essentially does what we talked about over discord, still getting function signature mismatch though, so I assume it's probably calling the wrong index in the table? :thinking: 

Also for monday, wanted to ask if these are the ids you are talking about: https://github.com/DuncanUszkay1/assemblyscript/blob/master/src/program.ts#L3479 :smile: (Asking here to not ping you when you aren't working, and remind myself later haha! :smile: )

Gonna work on this more Monday :smile:

### CODE REVIEW

File: src/compiler.ts [hunk 1]
```diff
@@ -8578,18 +8589,38 @@ export class Compiler extends DiagnosticEmitter {
...
+        console.log('torch2424 8615');
+
+        // Enter our closure function into our closures table
+        let index = this.ensureFunctionTableEntry(closureFunctionInstance);
+
+        // Set the param type of the function to the original function
+        this.currentType = originalFunctionInstance.signature.type;
+
+        // Return the closure function index
         return module.i32(index);
```
Comment (DuncanUszkay1) [inline src/compiler.ts]: We're still returning an index here, which is an invalid representation of a function. This has to be a pointer to a function struct like we do [here]( https://github.com/DuncanUszkay1/assemblyscript/blob/d69d861a8b7a4fe32e5f00908193e14c3541df4c/src/compiler.ts#L8335)

Comment (torch2424) [reply src/compiler.ts]: Ah! Thanks for catching that! Cool I'll go ahead and fix that :smile:

**Comment by torch2424** [comment]
Yay! Thank you for the help @DuncanUszkay1 !!! The test passes now :smile: :tada: 

Screenshot from 2020-06-22 18-17-59

I didn't have much time today for closures, but I wanted to try your idea, and it worked haha! :smile: 

Tommorrow I'm going to clean up this PR, and I think I'm going to move this logic into it's own function? As most of this is copy-paste form `compileFunctionExpression`, and I think it'd be better as a `createClosure` function or something like that :smile:

**Comment by DuncanUszkay1** [comment]
🎉 awesome, thanks. Have you had a chance to test that the correct values are being returned by these function calls? I've had some cases where tests were passing but the wrong values were being returned.

If not I might pick that up today since I have a whole setup for checking those values

**Comment by torch2424** [comment]
@DuncanUszkay1 This is now good to go! Cleaned everything up! :)

Screenshot from 2020-06-23 11-54-32

> Have you had a chance to test that the correct values are being returned by these function calls? I've had some cases where tests were passing but the wrong values were being returned.

I have the assertions here: https://github.com/DuncanUszkay1/assemblyscript/pull/4/files#diff-9823884c44802ad5e6c1e3ee07cfdb01R11

Unless you mean like memory values and things? Than no, but stoked to hear you have a method to check! I'd be interested in learning how to use it! :smile: :+1:

**Pull request merged**